Create rule S6953: Don't use `allowedValues` for a location parameter (#3796)
SonarSource/rspec: rules/S6953/azureresourcemanager/rule.adoc

> In Azure Resource Manager (ARM) templates, it is a common practice to set `allowedValues` for various parameters to limit the options and maintain control. However, when it comes to the "location" parameter, this practice can lead to a code smell. Users may be unable to deploy such a template if their desired location is not included in the `allowedValues`.

Reviewer: Is it really a common practice? Island AI also generated the same sentence for my rule and I don't trust it.

Reviewer 2: Yeah, I will change this to be more neutral

> In Azure Resource Manager (ARM) templates, it is a common practice to set `allowedValues` for various parameters to limit the options and maintain control. However, when it comes to the "location" parameter, this practice can lead to a code smell. Users may be unable to deploy such a template if their desired location is not included in the `allowedValues`.

Reviewer: You can use `location` instead of "location"

> In Azure Resource Manager (ARM) templates, it is a common practice to set `allowedValues` for various parameters to limit the options and maintain control. However, when it comes to the "location" parameter, this practice can lead to a code smell. Users may be unable to deploy such a template if their desired location is not included in the `allowedValues`.

Reviewer: Looking at it again, I don't like the duplication that much.
In the first section I would expect more an explanation what the allowedValues does.

> In Azure Resource Manager (ARM) templates, it is a common practice to set `allowedValues` for various parameters to limit the options and maintain control. However, when it comes to the "location" parameter, this practice can lead to a code smell. Users may be unable to deploy such a template if their desired location is not included in the `allowedValues`.

Reviewer: General comment. Please write one sentence per line https://asciidoctor.org/docs/asciidoc-recommended-practices/#one-sentence-per-line

== Why is this an issue?

> In Azure Resource Manager (ARM) templates, it is a common practice to set `allowedValues` for various parameters to limit the options and maintain control. However, when it comes to the "location" parameter, this practice can lead to a code smell. Specifically, setting `allowedValues` for a location parameter can cause issues because the locations you specify may not be available in all clouds. Users may be unable to deploy such a template if their desired location is not included in the `allowedValues`, causing inconvenience and potential delays in their work.

Reviewer: See above.

> In Azure Resource Manager (ARM) templates, it is a common practice to set `allowedValues` for various parameters to limit the options and maintain control. However, when it comes to the "location" parameter, this practice can lead to a code smell. Specifically, setting `allowedValues` for a location parameter can cause issues because the locations you specify may not be available in all clouds. Users may be unable to deploy such a template if their desired location is not included in the `allowedValues`, causing inconvenience and potential delays in their work.

Reviewer: In all cloud regions?

== How to fix it in ARM Templates

Remove `allowedValues` for the location parameter.

=== Code examples

==== Noncompliant code example

[source,json,diff-id=1,diff-type=noncompliant]
----
{
  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
  "contentVersion": "1.0.0.0",
  "parameters": {

>       "location": {

Reviewer: Do we want to include usage of this parameter? Or would it be too much then?

Reviewer 2: I think usage is not necessary in this case, as the rule is only about parameter specification

          },
          "defaultValue": "[resourceGroup().location]",
          "allowedValues": [
              "eastus",
              "westus",
              "northeurope",
              "westeurope",
              "southeastasia"
          ]
      }
  }
}
----

==== Compliant solution

[source,json,diff-id=1,diff-type=compliant]
----
{
  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
  "contentVersion": "1.0.0.0",
  "parameters": {
      "location": {
          "type": "string",
          "metadata": {
              "description": "The location in which the resources should be deployed."
          },
          "defaultValue": "[resourceGroup().location]"
      }
  }
}
----

== How to fix it in Bicep

Remove `allowedValues` for the location parameter.

=== Code examples

==== Noncompliant code example

[source,bicep,diff-id=2,diff-type=noncompliant]
----
param location string {
  allowed: [
    'eastus'
    'westus'
    'northeurope'
    'westeurope'
    'southeastasia'
  ]
}
----

==== Compliant solution

[source,bicep,diff-id=2,diff-type=compliant]
----

> param location string

Reviewer: I would make the example semantically the same as the above one, i.e. adding a default value.

----

== Resources
=== Documentation

> * https://learn.microsoft.com/en-us/azure/azure-resource-manager/templates/best-practices#location-recommendations-for-parameters/[Location recommendations for parameters - Azure Resource Manager]

Reviewer: This way the "jump" works correctly.
```suggestion
* https://learn.microsoft.com/en-us/azure/azure-resource-manager/templates/best-practices#location-recommendations-for-parameters[Location recommendations for parameters - Azure Resource Manager]
```

ifdef::env-github,rspecator-view[]

'''
== Implementation Specification
(visible only on this page)

=== Message
Remove this `allowedValues` property from the `location` parameter.

=== Highlighting

> Highlight the `allowedValues` property in the `location` parameter.

Reviewer: How should this be detected? Should this raise an issue if the parameter name equals "location" or if the parameter is used as an location in a resource? 
IMO we can't simply rely on the fact that the name of the parameter matches the intent, as users could name their parameter `virtualMachinesLocation`

Reviewer 2: I added my thoughts into the Jira ticket. I agree that the parameter name is not enough, I would suggest to look for a combination of a name and default value.

> In Azure Resource Manager (ARM) templates, it is possible to set `allowedValues` for various parameters.

Reviewer: I like the way how the description is now written!

This field is used to define a list of valid values for a parameter.
This is a way to restrict the input values provided when deploying or updating a resource using the template.

> However, when it comes to the `location` parameter, this practice can lead to a code smell.

Reviewer: I would word it like this, because there is no single location parameter
```suggestion
However, when it comes to a parameter defining the `location` of a resource, this practice can lead to a code smell.
```

Users may be unable to deploy such a template if their desired location is not included in the `allowedValues`.

== Why is this an issue?

In Azure Resource Manager (ARM) templates, it is possible to set `allowedValues` for various parameters to limit the options and maintain control.

> However, when it comes to the `location` parameter, this practice can lead to a code smell.

Reviewer: ```suggestion
However, when it comes to a parameter defining the `location` of a resource, this practice can lead to a code smell.
```

Specifically, setting `allowedValues` for a location parameter can cause issues because the locations list might not be exhaustive or suitable for all users.
Users may be unable to deploy such a template if their desired location is not included in the `allowedValues`, causing inconvenience and potential delays in their work.

== How to fix it in ARM Templates

> Remove `allowedValues` for the location parameter.

Reviewer: ```suggestion
Remove `allowedValues` for the parameter specifying the location.
```

=== Code examples

==== Noncompliant code example

[source,json,diff-id=1,diff-type=noncompliant]
----
{
  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
  "contentVersion": "1.0.0.0",
  "parameters": {
      "location": {
          "type": "string",
          "metadata": {
              "description": "The location in which the resources should be deployed."
          },
          "defaultValue": "[resourceGroup().location]",
          "allowedValues": [
              "eastus",
              "westus",
              "northeurope",
              "westeurope",
              "southeastasia"
          ]
      }
  }
}
----

==== Compliant solution

[source,json,diff-id=1,diff-type=compliant]
----
{
  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
  "contentVersion": "1.0.0.0",
  "parameters": {
      "location": {
          "type": "string",
          "metadata": {
              "description": "The location in which the resources should be deployed."
          },
          "defaultValue": "[resourceGroup().location]"
      }
  }
}
----

== How to fix it in Bicep

> Remove `allowedValues` for the location parameter.

Reviewer: ```suggestion
Remove `allowedValues` for the parameter specifying the location.
```

=== Code examples

==== Noncompliant code example

[source,bicep,diff-id=2,diff-type=noncompliant]
----
@allowed([
  'eastus'
  'westus'
  'northeurope'
  'westeurope'
  'southeastasia'
])  // Noncompliant

> param location string = resourceGroup().location

Reviewer: I see that in the Jira Ticket there is a different bicep example. Are both valid syntax? Then maybe I would include both in here and in the ticket.

Reviewer 2: No, I'll fix the example in Jira. It was hallucinated by Island AI; I didn't see any mentions that it is valid in Bicep.

----

==== Compliant solution

[source,bicep,diff-id=2,diff-type=compliant]
----
param location string = resourceGroup().location
----

== Resources

=== Documentation

* Azure Resource Manager - https://learn.microsoft.com/en-us/azure/azure-resource-manager/templates/best-practices#location-recommendations-for-parameters[Location recommendations for parameters]

ifdef::env-github,rspecator-view[]

'''
== Implementation Specification
(visible only on this page)

=== Message

> For ARM JSON: Remove this `allowedValues` property from the `location` parameter.

Reviewer: ```suggestion
For ARM Templates: Remove this `allowedValues` property from the parameter specifying the location.
```

> For Bicep: Remove this `@allowed` decorator from the `location` parameter.

Reviewer: ```suggestion
For Bicep: Remove this `@allowed` decorator from the parameter specifying the location.
```

=== Highlighting

> In case of AMR JSON, highlight the `allowedValues` property in the `location` parameter.

Reviewer: ```suggestion
In case of ARM Tempates , highlight the `allowedValues` property in the parameter specifying the location.
```

> In case of Bicep, highlight the `@allowed` decorator above the `location` parameter.

Reviewer: ```suggestion
In case of Bicep, highlight the `@allowed` decorator above the parameter specifying the location.
```
